Update mailings

mpacer opened the pull request:
So this does a few things to improve our existing email generating system.

The first thing it does is it standardises a few of the fields that had previously been template specific.

The second thing it does is introduces a new Mailer class which contains the common logic for all of the email scripts and reduces the amount of boilerplate we need to keep around in each script.

this new Mailer object is intended to be subclassed with the ability to customise the fields needed for the particular template with the `custom_data` property, as can be seen in the call_for_papers example.

@dalippa and @deniederhut If you have a chance to look at this, I'd appreciate it… i'm probably going to use this to send out the emails as soon as I have the list and I figure out how to hook this up to my gmail.

My (hard) todos on this are:

- [x] Get session to connect appropriately
- [x] Fix/clarify authentication logic
- [x] setup new authors json based on the papers info

My (soft) todos on this are:
- [ ] update other scripts to use Mailer
- [x] setup nicer addresses for the recipients

mpacer committed:
update email templates and email scripts to use new common idioms

mpacer committed:
update email.json metadata to be mostly up-to-date

mpacer committed:
update mailer to provide common routines for creating common info

mpacer committed:
introduce new Mailer class

also adds docstring and appropriate imports

mpacer committed:
add new call_for_papers email template

mpacer commented:
@gforsyth this is the email system we currently have up and running, if you have any suggestions on how to adapt it for the json that you'll be creating I'd appreciate it!

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -46,25 +50,67 @@ def get_password(sender):
+    if data is None:
+        data = cfg2dict(proc_conf)
+    proc = data.get('proceedings', {})
+    editors = proc.get('editor', [])
```
to be consistent, consider using editor or editors, not both

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -46,25 +50,67 @@ def get_password(sender):
     return '"%s" <%s>' % (name_email['name'], name_email['email'])
 
 
+def create_committee(data=None):
```
Why not use `**kwargs` here?

mpacer replied:
that might be possible… but I generally try to avoid using **kwargs and then testing if it's none. But I see where you're coming from.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -46,25 +50,67 @@ def get_password(sender):
         print(message)
         print("=" * 80)
-
         return
```
Consider moving the return statement below the `smtplib.SMTP` call so you can use the dry run to verify credentials against the mail server.

mpacer replied:
I got this from the old way of approaching the problem… we could do that, but it would require passing in a password each time you wanted to do the dry_run, which would have put a huge amount of overhead when trying to iterate quickly over this code.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -76,3 +122,170 @@ def send_template(sender, recipient, template, template_data,
+        if os.path.exists(source_file):
+            return cfg2dict(source_file)
+        else:
+             print('file at {} not found'.format(os.path.abspath(s)))
```
If the data file doesn't exist, would it be better to raise an exception since the dependent functions cannot do what they need to do otherwise?

mpacer replied:
yes, I'll make that happen.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -76,3 +122,170 @@ def send_template(sender, recipient, template, template_data,
+             print('file at {} not found'.format(os.path.abspath(s)))
+
+
+def load_data_sources(sources):
```
I think there is a way to merge dictionaries with `x.update(y)`. Why not use that instead of creating the new_dict over and over?

mpacer replied:
You're right, there's gotta be a better way than this… i was trying to avoid update because of the way it changes everything in place, but you can't unpack dictionaries in a comprehension as far as I could tell.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -46,25 +50,67 @@ def get_password(sender):
+            for name, email 
+            in zip(editors, editor_email)]
+
+def editor_email_string(data=None):
```
Minor comment: these methods are implementation details used by the class; from a design standpoint, prefixing "private" methods with `_` is the usual convention.

mpacer replied:
I moved these methods inside the class, and now they should be able to be overwritten if needed… whether that would be necessary I'm not sure, but I can make them private if you want.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -76,3 +122,170 @@ def send_template(sender, recipient, template, template_data,
+
+    @property
+    def recipients(self):
+        return 'blah@blah.org'
```
Is this a comma or semicolon delimited list? Usually a property that is in plural form returns a list, not a single string. At a minimum, consider documenting so classes that override this method would know the contract.

mpacer replied:
Honestly, I was treating this as a string because that's sort of how it was being done before. 

In any case, I was slightly inclined to raise a `NotImplementedError`… since it shouldn't need to be set unless someone wants to do so explicitly.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -76,3 +122,170 @@ def send_template(sender, recipient, template, template_data,
+        
+        self.data_sources = ['./email.json'] + data_sources
+        
+        self.dry_run = True
```
You use `self._sender` for member variables later, but `self.dry_run` here. Consider a consistent naming convention.

mpacer replied:
I was using underscores to indicate private member variables not just member variables in general. I usually do that whenever I have a property for the public member variable, but still need to store the info somewhere.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -76,3 +122,170 @@ def send_template(sender, recipient, template, template_data,
+        
+    @property
+    def sender(self):
+        return self._sender if self._sender else self.template_data["sender"]
```
What if the template_data has an empty string or something? Consider `self.template_data.get("sender")` which would raise an exception or `self.template_data.get("sender", "cochairs@scipy.org")` or something like that to provide a default.

mpacer replied:
If we use `self.template_data.get("sender")` it would return None, which wouldn't return an exception locally, it would raise an exception somewhere further down the line. 

also sender looks to be expected as a dict not a string, so I should probably make that work consistently.

mpacer replied:
This is making me realise just how much of this logic is only for validating and working with these "name" "email" pairs… I should probably factor that out into its own class…

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -76,3 +122,170 @@ def send_template(sender, recipient, template, template_data,
+            self.send_mail(recipients, message)
+
+
+    def display_message(self, recipients, message):
```
maybe `display_dryrun_message`?

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -46,25 +50,67 @@ def get_password(sender):
+    
+    message = create_message(recipient, template, template_data)
 
     if args.dry_run:
```
consider making the dry run message in the class below a static method and call it here.

mpacer replied:
Great idea… I'll implement that shortly.

dalippa commented on publisher/mail/email.json:
```diff
@@ -1,22 +1,24 @@
-   "built_proceedings_url": "http://jarrodmillman.com/scipy2011",
+   "built_proceedings_url": "http://procbuild.scipy.org",
 
    "reviewers": [
```
Change the reviewers too?

mpacer replied:
They're just test reviewers… it actually helped me realise that because they were from so long ago.

dalippa commented on publisher/mail/templates/call_for_papers.txt.tmpl:
```diff
@@ -0,0 +1,46 @@
+Dear {{author}},
+
+Congratulations on having your abstract accepted for this year's event -- we 
+look forward to seeing your poster or presentation! As in the past, SciPy2018 
```
Use `{{year}}` here as well

mpacer replied:
I think this might be `{{conference}}`, but definitely a good idea.

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -76,3 +122,170 @@ def send_template(sender, recipient, template, template_data,
 
     session.sendmail(sender['name'], recipient, message)
     session.quit()
```
try/finally on session quit?

mpacer replied:
I'll add this in both here and the method.

mpacer replied:
actually… doesn't the context manager already handle this?

mpacer committed:
create fancy_prep and fancy_emails methods for handling author/email info

mpacer committed:
create class methods for editor_email_string and create_committee

these now use the fancy_prep and fancy_emails methods.

mpacer committed:
update call_for_papers to use new _mailer logic, use new accepted… json

mpacer committed:
update template to use conference where we say scipy2018

dalippa commented on publisher/mail/_mailer.py:
```diff
@@ -175,27 +175,60 @@ def __init__(self,
+            this is the key to indicate emails
+        """
+        people_gen = (p for p in cls.fancy_prep(data, name_key=name_key, email_key=email_key))
+        are_str = lambda x: (isinstance(x['name'], str) and isinstance(x['email'], str)) 
```
not sure what `are_str` is here. Do you mean `is_str`?

dalippa commented on publisher/mail/mail_call_for_papers.py:
```diff
@@ -2,30 +2,32 @@
     @property
     def recipients(self):
-        return ', '.join(self.template_data['author_email'])
+        # this is for testing
```
Minor: Consider moving this into a test file and creating a "fake" class `TestAuthorMailer` that subclasses `AuthorMailer` and does this.

dalippa commented on publisher/mail/templates/call_for_papers.txt.tmpl:
```diff
@@ -22,8 +22,11 @@ scientific computing in Python, such as:
-point, the six-week open review period will begin. During this time, members
+To submit a paper to the proceedings, submit a pull request to {{github_url}}.
+
+Papers must be submitted as pull requests by {{submission_date}}. 
```
Consider combining this sentence with the prior and following sentences:

"To submit a paper to the proceedings, submit a pull request to {{github_url}} by {{submission_date}}, when the six-week open review period begins."

mpacer replied:
I originally did it and it read rather long & not obviously containing crucial information. These are both important pieces of information so I wanted them to be highlighted separately.

dalippa reviewed (approved):
Minor recommendations. Feel free to fix them before merging, or after.

mpacer committed:
minor fixes and repsonses to dalippa's PR comments

mpacer committed:
changes made for getting emails to be sent to more than one person

also small changes to make sure it was targeting my correct account &
for debugging, should be cleaned up later.

mpacer committed:
separate logic for writing cc headers (string) vs sending via smtp(list)

mpacer committed:
move AuthorMailer out of loop to reuse credentials; use fancy_email_list

mpacer committed:
Encode utf8; use temp_data; Greet 2 authors; title in template

remove debug statements

mpacer commented:
This was able to send off most of the emails we needed to send. With minimal issues.

For some reason it swallowed the `+scipyproc` from mpacer.phd+scipyproc@gmail.com, not sure why.

mpacer commented:
Ah — I think it it probably swallowed that because of the "login" value. I'll play around with that a bit to see if adding `.` and `+*` works as a login value.
